src/ros_filter.cpp

```diff
@@ -527,20 +533,67 @@ namespace RobotLocalization
           restoredMeasurementCount--;
         }
 
+        MeasurementPtr measToProcess = measurement;
+
+        if (measurement->differential_)
+        {
+          MeasurementPtr differentialMeas = boost::make_shared<Measurement>();
+          *differentialMeas = *measurement;
+
+          applyPreviousState(differentialMeas);
+
+          measToProcess = differentialMeas;
+        }
+
         // This will call predict and, if necessary, correct
-        filter_.processMeasurement(*(measurement.get()));
+        filter_.processMeasurement(*(measToProcess.get()));
+
+        // Accumulate all differential topics that have this timestamp
+        if (measurement->differential_)
+        {
+          differentialTopicsToSave.push(measurement->topicName_);
+        }
 
         // Store old states and measurements if we're smoothing
         if (smoothLaggedData_)
         {
           // Invariant still holds: measurementHistoryDeque_.back().time_ < measurementQueue_.top().time_
           measurementHistory_.push_back(measurement);
+        }
+
+        // We should only save the filter state once per unique timstamp/topic combo
+        if (measurementQueue_.empty() ||
+          ::fabs(measurementQueue_.top()->time_ - filter_.getLastMeasurementTime()) > 1e-9)
+        {
+          bool curMeasIsDifferential = false;
 
-          // We should only save the filter state once per unique timstamp
-          if (measurementQueue_.empty() ||
-              ::fabs(measurementQueue_.top()->time_ - filter_.getLastMeasurementTime()) > 1e-9)
+          // The time stamp of the next measurement is going to change (or we have no more). That means that, at this
+          // point, differentialTopicsToSave contains the name of all differential topics with measurements we received
+          // with this exact timestamp. We want them all to get associated with the most recent corrected state, so now
+          // we need to walk through each of them and do so.
+          while (!differentialTopicsToSave.empty())
           {
-            saveFilterState(filter_);
+            if (smoothLaggedData_)
+            {
+              // If we're smoothing, we *always* save the filter state and associate it with the most recent
+              // measurement (after this while loop), so we can skip that one here.
+              saveFilterState(differentialTopicsToSave.front());
+              previousMeasurementStates_[differentialTopicsToSave.front()] = filterStateHistory_.back();
+            }
+            else
+            {
+              // If we're not smoothing, we still need to keep track of the previous state for this sensor
+              previousMeasurementStates_[differentialTopicsToSave.front()] =
+                generateFilterState(differentialTopicsToSave.front());
+            }
+
+            differentialTopicsToSave.pop();
+          }
+
+          // Only save the state if we didn't already do so above
```

Comment:
The logic might be correct, but it does not match exactly match the comment. For example, if `smoothLaggedData_` is false, `saveFilterState` will not be called.

Reply:
Ah, true. Yes, the logic is as I want it to be, but the comment isnot clear.